### PR TITLE
[qa] Ensure bitcoind processes are cleaned up when tests end

### PR DESCRIPTION
**Pull request by sdaftuar**
When tests fail (such as due to a bug in the test, race condition, etc), it's possible that we could follow code paths that bypass our normal node shutdown that occurs in `TestNode.stop_node`.  Add a destructor to `TestNode` that cleans this up.

### CODE REVIEW

**Comment by sdaftuar** [comment]
I have been able to reproduce the issue of bitcoind processes living on after tests exit in a couple ways:
 * #12375: running wallet_encryption.py on macos fails and leaves a bitcoind running
 * the feature_dbcrash.py failure that should be fixed in #12902 does the same.

This PR causes all nodes to exit in both those situations.

**Comment by MarcoFalke** [comment]
utACK 9caecce6c7b00d3cc43eb220e91e9d1ac42251e8, makes sense

**Comment by laanwj** [comment]
Should this heed --noshutdown?

File: test/functional/test_framework/test_node.py [hunk 1]
```diff
@@ -84,6 +84,13 @@ def __init__(self, i, datadir, rpchost, timewait, binary, stderr, mocktime, cove
 
         self.p2ps = []
 
+    def __del__(self):
+        # Ensure that we don't leave any bitcoind processes lying around after
+        # the test ends
+        if self.process:
+            self.log.info("Cleaning up leftover process")
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py]: Looks like the log is flushed and closed when the tests are done, so wouldn't work at this point.

Comment (sdaftuar) [reply test/functional/test_framework/test_node.py]: Odd, I was able to see this logging work when I tested locally.  Will remove.

**Comment by sdaftuar** [comment]
> Should this heed --noshutdown?

@laanwj Oops, yes of course it should, will fix.

File: test/functional/test_framework/test_node.py (hunk 1 above)
Comment (jnewbery) [reply test/functional/test_framework/test_node.py]: Yes, it appears that logging *can* work after calling `logging.shutdown()`, but the documentation states that no calls to logging *should* be made after shutdown: https://docs.python.org/3/library/logging.html#logging.shutdown

**Comment by jnewbery** [comment]
Seems reasonable. I don't think this can do any harm.

utACK ccc9031c69e67c336362b197c7decf14af98a877

**Comment by MarcoFalke** [comment]
Might have to set `self.process` to `None` in `feature_help.py`?

**Comment by promag** [comment]
Tested ACK f1d91ef.

Using the instructions in https://github.com/bitcoin/bitcoin/pull/12902#issuecomment-379260451 with this change cleans kills all test processes.

Code looks good, please squash.

**Comment by jonasschnelli** [comment]
utACK f1d91ef5a2e91c07060b39f0285a733ecf6831b3

**Comment by laanwj** [comment]
utACK e36a0c08529bccc695ec71a7ec1df89367cc1628

**Pull request merged**